### PR TITLE
Fix the discriminator property inheritance issue

### PR DESCRIPTION
**Pull request by ArcturusZhang**
Fixes https://github.com/Azure/autorest.csharp/issues/4092
Depending on https://github.com/Azure/autorest.csharp/pull/4124

# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Output/Models/Serialization/Json/JsonAdditionalPropertiesSerialization.cs [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 namespace AutoRest.CSharp.Output.Models.Serialization.Json
 {
-    internal class JsonAdditionalPropertiesSerialization : JsonPropertySerialization
+    internal record JsonAdditionalPropertiesSerialization : JsonPropertySerialization
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Output/Models/Serialization/Json/JsonAdditionalPropertiesSerialization.cs:11]: changing this to record does not really change anything in the runtime's perspective.
But record has its own "ToString" implementation, which simplifies the debugging process.
Therefore I leave these changes in this PR.

File: src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs [hunk 2]
```diff
@@ -155,7 +156,7 @@ private ModelTypeProviderFields EnsureFields()
         private ConstructorSignature EnsurePublicConstructorSignature()
         {
             //get base public ctor params
-            GetConstructorParameters(Fields.PublicConstructorParameters, out var fullParameterList, out var parametersToPassToBase, out var baseInitializers, true);
+            GetConstructorParameters(Fields.PublicConstructorParameters, out var fullParameterList, out var baseInitializers, true);
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:159]: last time I change this method, this output parameter is not used anywhere but I forgot to remove it. Here we just remove it

File: src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs [hunk 3]
```diff
@@ -283,11 +303,29 @@ private void GetConstructorParameters(IEnumerable<Parameter> parameters, out IRe
...
             }
             parameterList.AddRange(parameters);
-            fullParameterList = parameterList;
+            fullParameterList = parameterList.DistinctBy(p => p.Name).ToArray(); // we filter out the parameters with the same name since we might have the same property both in myself and my base.
+        }
+
+        private bool? _isDiscriminatorInheritedOnBase;
+        internal bool IsDiscriminatorInheritedOnBase => _isDiscriminatorInheritedOnBase ??= EnsureIsDiscriminatorInheritedOnBase();
+
+        private bool EnsureIsDiscriminatorInheritedOnBase()
```
Comment (chunyu3) [inline src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:315]: Just want to confirm if we need to consider override, as following, and what's the expected behavior?
```
model base {
  kind: string;
}

@discriminator("kind")
model baseWithDiscriminator extends base {
  kind: string,
  name: string;
}

model derived extends baseWithDiscriminator {
  kind: "A",
  age: int32;
}
```

File: src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs [hunk 4]
```diff
@@ -399,10 +438,25 @@ private IReadOnlyList<ObjectPropertyInitializer> GetPropertyInitializers(IReadOn
...
+        }
+
         protected override IEnumerable<ObjectTypeProperty> BuildProperties()
         {
+            var existingPropertyNames = GetParentPropertyNames();
             foreach (var field in Fields)
-                yield return new ObjectTypeProperty(field, Fields.GetInputByField(field));
+            {
+                var property = new ObjectTypeProperty(field, Fields.GetInputByField(field));
+                if (existingPropertyNames.Contains(property.Declaration.Name))
```
Comment (lirenhe) [inline src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:456]: To confirm, if one property whose name also exists in both ParentPropertyNames but with different type, which one will you use?

File: src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs [hunk 5]
```diff
@@ -573,6 +625,18 @@ private bool TryGetPropertyForInputModelProperty(Func<ObjectTypeProperty, bool>
             );
         }
 
+        private ObjectTypeProperty GetPropertyForDiscriminator(string inputPropertyName)
+        {
+            foreach (var obj in EnumerateHierarchy())
+            {
+                var property = obj.Properties.FirstOrDefault(p => p.InputModelProperty?.Name == inputPropertyName);
```
Comment (lirenhe) [inline src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:632]: will this ensure we pick the right one if there is a duplicate name of the properties for the discriminator?

File: src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs (hunk 3 above)
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:315]: we have exactly the same case in the `Models-TypeSpec` case.

File: src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs (hunk 4 above)
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:456]: we use the inherited one.
Overriding a property and changing its type (type narrow down) is not supported.
C# does not support type narrow down, therefore to support it, we need to design what does it look like first.
Issue tracking that: https://github.com/Azure/autorest.csharp/issues/4096
Now we just ignore all attempts to override the property in base.

For example:
```
model Base {
	a: string | int;
}

model Derived extends Base {
	a: string;
}
```
We will ignore the `a` defined in `Derived` (this piece of code did this).

When we need to support [type narrow down](https://github.com/Azure/autorest.csharp/issues/4096), we will go over the design first, and then change this part to implement it.

File: src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs (hunk 5 above)
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:632]: This is filtering through `Properties` on `ObjectType`, which we have already filtered the properties with duplicated names (see [this comment](https://github.com/Azure/autorest.csharp/pull/4097#discussion_r1449859334))

**Review by lirenhe — APPROVED** [review]

File: src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs (hunk 3 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProvider.cs:315]: @ArcturusZhang  The one you added is a normal case, that the `kind` is in the parent of the base model, but the base model does not override `kind` property.

**Pull request merged**